mllib/src/main/scala/org/apache/spark/ml/feature/OneHotEncoderEstimator.scala

```diff
@@ -0,0 +1,439 @@
+
+    val outputFields = inputColNames.zip(outputColNames).map { case (inputColName, outputColName) =>
+
+      require(schema(inputColName).dataType.isInstanceOf[NumericType],
+        s"Input column must be of type NumericType but got ${schema(inputColName).dataType}")
+      require(!inputFields.exists(_.name == outputColName),
+        s"Output column $outputColName already exists.")
+
+      OneHotEncoderCommon.transformOutputColumnSchema(
+        schema(inputColName), $(dropLast), outputColName)
+    }
+    StructType(inputFields ++ outputFields)
+  }
+
+  @Since("2.3.0")
+  override def fit(dataset: Dataset[_]): OneHotEncoderModel = {
+    val transformedSchema = transformSchema(dataset.schema)
+
+    val categorySizes = $(outputCols).zipWithIndex.map { case (outputColName, idx) =>
+      val outputAttrGroupFromSchema = AttributeGroup.fromStructField(
+        transformedSchema(outputColName))
+
+      val outputAttrGroup = if (outputAttrGroupFromSchema.size < 0) {
+        OneHotEncoderCommon.getOutputAttrGroupFromData(
+          dataset, $(dropLast), $(inputCols)(idx), outputColName)
+      } else {
+        outputAttrGroupFromSchema
+      }
+
+      outputAttrGroup.size
+    }
+
+    val model = new OneHotEncoderModel(uid, categorySizes).setParent(this)
+    copyValues(model)
+  }
+
+  @Since("2.3.0")
+  override def copy(extra: ParamMap): OneHotEncoderEstimator = defaultCopy(extra)
+}
+
+@Since("2.3.0")
+object OneHotEncoderEstimator extends DefaultParamsReadable[OneHotEncoderEstimator] {
+
+  private[feature] val SKIP_INVALID: String = "skip"
+  private[feature] val ERROR_INVALID: String = "error"
+  private[feature] val supportedHandleInvalids: Array[String] = Array(SKIP_INVALID, ERROR_INVALID)
+
+  @Since("2.3.0")
+  override def load(path: String): OneHotEncoderEstimator = super.load(path)
+}
+
+@Since("2.3.0")
+class OneHotEncoderModel private[ml] (
+    @Since("2.3.0") override val uid: String,
+    @Since("2.3.0") val categorySizes: Array[Int])
+  extends Model[OneHotEncoderModel] with OneHotEncoderParams with MLWritable {
+
+  import OneHotEncoderModel._
+
+  private def encoders: Array[UserDefinedFunction] = {
+    val oneValue = Array(1.0)
+    val emptyValues = Array.empty[Double]
+    val emptyIndices = Array.empty[Int]
+    val dropLast = getDropLast
+    val handleInvalid = getHandleInvalid
+
+    categorySizes.map { size =>
+      udf { label: Double =>
+        if (label < size) {
+          Vectors.sparse(size, Array(label.toInt), oneValue)
+        } else if (label == size && dropLast) {
+          Vectors.sparse(size, emptyIndices, emptyValues)
+        } else {
+          if (handleInvalid == OneHotEncoderEstimator.ERROR_INVALID) {
+            throw new SparkException(s"Unseen value: $label. To handle unseen values, " +
+              s"set Param handleInvalid to ${OneHotEncoderEstimator.SKIP_INVALID}.")
+          } else {
+            Vectors.sparse(size, emptyIndices, emptyValues)
```

Comment:
This behaviour seems more like "keep" in StringIndexer than "skip". Skip filters out rows with invalid values and returns a dataframe with number of rows <= the number of rows in the input. "keep" maps all invalid values to a single, known value. Should we rename this to `KEEP_INVALID`?

https://spark.apache.org/docs/2.2.0/ml-features.html#stringindexer

Reply:
When I checked scikit-learn's OneHotEncoder: http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html, its handle_unknown has `error` or `ignore` choices. Sounds like `ignore` is more close to `skip` in semantics?

I'm willing to rename it to `keep` if all you think it is better.

Reply:
Although `keep` does seem a little strange semantically, it's probably best to match the param option names from `StringIndexer` since that will be the typical pipeline `StringIndexer` -> `OneHotEncoder`

Reply:
Ok. Then I will rename it to `keep`. Thanks.